edge-hub/src/Microsoft.Azure.Devices.Routing.Core/Dispatcher.cs

```diff
@@ -21,19 +20,19 @@ public class Dispatcher : IDisposable
         readonly CancellationTokenSource cts;
         readonly IEndpointExecutorFactory endpointExecutorFactory;
         readonly AtomicReference<ImmutableDictionary<string, IEndpointExecutor>> executors;
-        readonly ICheckpointer checkpointer;
+        readonly MasterCheckpointer masterCheckpointer;
```

Comment:
What's the idea of the MasterCheckpointer and why are we using it here?

Reply:
this has always been an instance of MasterCheckpointer, it just used to be cached in two places:
1. here as the "global" checkpointer to checkpoint orphaned messages
2. in CheckpointerEndpointExecutorFactory as an instance of ICheckpointerFactory

Now that the executor is responsible for creating the checkpointer, CheckpointerEndpointExecutorFactory is not useful anymore so the ICheckpointerFactory reference is gone.  I just changed the type here to the object-class so that we can access both ICheckpointer and ICheckpointerFactory interfaces as needed.

---
In reply to: [390015947](https://github.com/Azure/iotedge/pull/2670#discussion_r390015947) [](ancestors = 390015947)

Reply:
where do you need to change type from ICheckpointer to MasterCheckpointer? MasterCheckpointer implements both ICheckpointer and ICheckpointerFactory.  Do you need to use ICheckpointerFactory?  if not, ICheckpointer should be used same as before. Do I miss anything?

Reply:
the operative change here is to pass in ICheckpointerFactory to the executor, because the executor now needs to create a different checkpointer per priority.  So we need ICheckpointerFactory everytime the Dispatcher creates a new executor, both for startup and for "on-the-fly" route changes.

---
In reply to: [390535732](https://github.com/Azure/iotedge/pull/2670#discussion_r390535732) [](ancestors = 390535732)

Reply:
Yeah, I think its safe to make the interface use the concrete `MasterCheckpointer`,